tokio/src/runtime/tests/loom_blocking.rs

```diff
@@ -23,6 +23,55 @@ fn blocking_shutdown() {
     });
 }
 
+#[test]
+fn spawn_mandatory_blocking_should_always_run() {
+    use crate::runtime::tests::loom_oneshot;
+    loom::model(|| {
+        let rt = runtime::Builder::new_current_thread().build().unwrap();
+
+        let (tx, rx) = loom_oneshot::channel();
+        let _enter = rt.enter();
+        runtime::spawn_blocking(|| {});
+        runtime::spawn_mandatory_blocking(move || {
+            let _ = tx.send(());
+        })
+        .unwrap();
+
+        drop(rt);
+
+        // This call will deadlock if `spawn_mandatory_blocking` doesn't run.
+        let () = rx.recv();
+    });
+}
+
+#[test]
+fn spawn_mandatory_blocking_should_run_even_when_shutting_down_from_other_thread() {
+    use crate::runtime::tests::loom_oneshot;
+    loom::model(|| {
+        let rt = runtime::Builder::new_current_thread().build().unwrap();
+        let _enter = rt.enter();
+
+        // Drop the runtime in a different thread
+        {
+            loom::thread::spawn(move || {
+                drop(rt);
+            });
```

Comment:
I don't understand how this can compile since the `EnterGuard` borrows the runtime object.

Reply:
Hmmmm interesting... Maybe the guard is getting dropped before moving the runtime... which misses the point of using it

Reply:
Yes, well, it shouldn't compile in the first place.

Reply:
I am now very puzzled by this as well. I thought the `_enter` was getting dropped before constructing the closure so that `rt` could be moved but that's not what's happening. If I call `drop(_enter);` explicitly before `thread::spawn`, the test fails at runtime because the context is not set when calling to `spawn_mandatory_blocking`. However, if I call `drop(_enter)` explicitly after `spawn`, then that doesn't compile because `rt` is moved and so the borrow is not valid anymore. If none of these are ok, where is the compiler inserting the `drop`? why does that compile? Could this be a bug in the borrow checker? Or do I just not understand how the borrow checker works?

Reply:
I believe it has to do with the thing described on [this page][1]. It relies on the fact that the guard doesn't have a destructor (it has a field with a destructor, but that field has no lifetime).

[1]: https://doc.rust-lang.org/nomicon/dropck.html

Reply:
You are correct, if I add a (trivial) destructor to `EnterGuard`, then the test doesn't compile anymore. 

I'm wondering how I could write the test differently so that this isn't a problem. Or is it fine to keep it like this? I'll give it a thought and see if I can write the test in a way that doesn't rely on `EnterGuard` not hitting the drop check

Reply:
You can use a `Handle` to the runtime.

Reply:
Ah of course,  I was under the impression that getting a handle requires borrowing the runtime but forgot that the handle can be cloned.